app/models/assignment.rb

```diff
@@ -392,25 +392,71 @@ def add_group(new_group_name = nil)
     Grouping.create(group: group, assignment: self)
   end
 
-  def add_group_api(new_group_name = nil)
-    if new_group_name.nil?
-      if group_name_autogenerated
-        group = self.course.groups.new
-        group.save(validate: false)
-        group.group_name = group.get_autogenerated_group_name
-        group.save
-      else
-        raise 'A group name was not provided'
+  def add_group_api(new_group_name = nil, members = [])
+    members ||= []
+    group = nil
+
+    Group.transaction do
+      group = if new_group_name.nil?
+                if members.length == 1 && self.group_max == 1
+                  student_user_name = members.first
+                  if self.is_timed
+                    group = course.groups.new
+                    group.save(validate: false)
+                    group.group_name = group.get_autogenerated_group_name
+                    group
+                  else
+                    Group.find_or_initialize_by(group_name: student_user_name, course: self.course) do |g|
+                      g.repo_name = student_user_name
+                    end
+                  end
+                elsif group_name_autogenerated
+                  group = course.groups.new
+                  group.save(validate: false)
+                  group.group_name = group.get_autogenerated_group_name
+                  group
+                else
+                  raise 'A group name was not provided'
+                end
+              else
+                existing_group = self.course.groups.find_by(group_name: new_group_name)
+                if existing_group
+                  raise "Group '#{new_group_name}' already exists" if groupings.exists?(group_id: existing_group.id)
+                  existing_group
+                else
+                  Group.new(group_name: new_group_name, course: self.course, repo_name: new_group_name)
+                end
+              end
+
+      group.save!
+
+      if groupings.exists?(group_id: group.id)
+        raise "Group '#{group.group_name}' is already part of this assignment"
       end
-    elsif (group = self.course.groups.where(group_name: new_group_name).first)
-      unless groupings.where(group_id: group.id).first.nil?
-        raise "Group #{new_group_name} already exists"
+      grouping = Grouping.create!(group: group, assignment: self)
+
+      unless members.empty?
+        students = Student.joins(:user)
+                          .where(users: { user_name: members })
+                          .index_by { |s| s.user.user_name }
```

Comment:
This is a good time to use `pluck`, as you really only need the user name and id here